### PR TITLE
Run all browser tests in 2gb mode

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

File: test/test_browser.py [hunk 1]
```diff
@@ -237,6 +237,8 @@ def setUp(self):
   def require_wasm2js(self):
     if self.is_wasm64():
       self.skipTest('wasm2js is not compatible with MEMORY64')
+    if self.is_2gb() or self.is_4gb():
+      self.skipTest('wasm2j does not support over 2gb of memory')
```
Comment (dschuff) [inline test/test_browser.py:241]:
```suggestion
      self.skipTest('wasm2js does not support over 2gb of memory')
```

File: test/test_browser.py [hunk 2]
```diff
@@ -2712,7 +2725,7 @@ def test_html5_webgl_create_context_no_antialias(self, args):
     '': ([],),
     'closure': (['-O2', '-g1', '--closure=1'],),
     'full_es2': (['-sFULL_ES2'],),
-    'pthread': (['-pthread'],),
+    'pthread': (['-pthread', '-sGL_DEBUG'],),
```
Comment (dschuff) [inline test/test_browser.py:2728]: I assume this was for debugging, do we still want it on?

**Review by dschuff — APPROVED** [review]

**Comment by sbc100** [comment]
ptal, which change is now fairly minimal as almost all substantive changes have been split out.

**Review by dschuff — APPROVED** [review]

File: src/library_html5_webgl.js [hunk 3]
```diff
@@ -14,8 +14,9 @@ var LibraryHtml5WebGL = {
     var len = arr.length;
     var writeLength = dstLength < len ? dstLength : len;
     var heap = heapType ? HEAPF32 : HEAP32;
+    dst = {{{ getHeapOffset('dst', 'float') }}};
```
Comment (kripken) [inline src/library_html5_webgl.js:17]: I guess this works because the offset is the same for float and i32 but maybe worth a comment?

File: test/test_browser.py [hunk 4]
```diff
@@ -3432,6 +3442,8 @@ def test_sdl2_mixer_music(self, formats, flags, music_name):
       '-sUSE_SDL=2',
       '-sUSE_SDL_MIXER=2',
       '-sSDL2_MIXER_FORMATS=' + ','.join(formats),
+      '-lc++',
+      '-lc++abi',
```
Comment (kripken) [inline test/test_browser.py:3446]: Why are these needed?

File: test/test_browser.py [hunk 5]
```diff
@@ -5910,14 +5963,16 @@ def setUp(self):
...
+    # Without this we get a warning about GLOBAL_BASE being ignored when used with SIDE_MODULE
+    self.emcc_args.append('-Wno-unused-command-line-argument')
     self.require_wasm64()
 
 
 class browser64_2gb(browser):
   def setUp(self):
     super().setUp()
     self.set_setting('MEMORY64')
-    self.set_setting('INITIAL_MEMORY', '2200gb')
```
Comment (kripken) [inline test/test_browser.py:5920]: Wow this was trying to use a lot of memory before :smile: how did it not error?

Comment (sbc100) [reply test/test_browser.py:5920]: We don't run any tests in this mode.  We normally use the `browser64_4gb` mode.. this mode was added for local testing

File: test/test_browser.py (hunk 4 above)
Comment (sbc100) [reply test/test_browser.py:3446]: Because libmodplug is written in C++.  I've added a comment and limited it to just that mode.

File: test/test_browser.py [hunk 6]
```diff
@@ -3425,14 +3435,19 @@ def test_sdl2_mixer_wav(self, flags):
...
       '-DFLAGS=' + flags,
       '-sUSE_SDL=2',
       '-sUSE_SDL_MIXER=2',
       '-sSDL2_MIXER_FORMATS=' + ','.join(formats),
-    ])
+    ]
+    # libmodplug is written in C++ so we need to link in C++
+    # libraries when using it.
+    if 'mod' in formats:
+      args += ['-lc++', '-lc++abi']
```
Comment (kripken) [inline test/test_browser.py:3449]: Would it be simpler maybe to make this a cpp test rather than c? I don't feel strongly though. I guess this is the workaround a C programmer would need, anyhow.

**Review by kripken — APPROVED** [review]

File: test/test_browser.py (hunk 6 above)
Comment (sbc100) [reply test/test_browser.py:3449]: I think we its nice to make sure that libsdl-mixer can be used from C programs.

As you know I really like not have C++ in the mix when debugging (unless the specific test needs C++).   Especially for debugging the linker its much easier not to have to see (the very long) C++ symbols in logging/tracing.
